tests/framework/prototype.py

```diff
@@ -0,0 +1,247 @@
+import yaml
+import copy
+from typing import Optional
+from kubernetes import client, config, utils
+import os
+import time
+import docker
+import logging
+import unittest
+
+# Utility functions
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+def search_path(cr, steps):
+    curr = cr
+    for step in steps:
+        if step.isnumeric():
+            int_step = int(step)
+            if int_step >= len(curr) or int_step < 0:
+                return None
+            curr = curr[int(step)]
+        elif step in curr:
+            curr = curr[step]
+        else:
+            return None
+    return curr
+
+def yaml_to_file(cr_yaml, fn):
+    f = open(fn, 'w')
+    f.write(yaml.dump(cr_yaml))
+    f.close()
+'''
+Functions for cluster preparation. Typical workflow:
+  Delete KinD cluster -> Create KinD cluster -> Install CRD -> Download Images (from DockerHub) ->
+  Load images into KinD cluster -> Install KubeRay operator
+'''
+def delete_kind_cluster():
+    os.system("kind delete cluster")
+
+def create_kind_cluster():
+    os.system("kind create cluster")
+    os.system("kubectl wait --for=condition=ready pod -n kube-system --all --timeout=900s")
+
+def install_crd():
+    KUBERAY_VERSION = "v0.3.0"
+    os.system("kubectl create -k \"github.com/ray-project/kuberay/manifests/cluster-scope-resources?" +
+              "ref={}&timeout=90s\"".format(KUBERAY_VERSION))
+
+def download_images(images):
+    client = docker.from_env()
+    for image in images:
+        client.images.pull(image)
+    client.close()
+
+def kind_load_images(images):
+    for image in images:
+        os.system('kind load docker-image {}'.format(image))
+
+def install_kuberay_operator():
+    KUBERAY_VERSION = "v0.3.0"
+    os.system("kubectl apply -k \"github.com/ray-project/kuberay/manifests/base?" +
+              "ref={}&timeout=90s\"".format(KUBERAY_VERSION))
+    time.sleep(60)
+
+'''
+Configuration Test Framework Abstractions: (1) DeltaSet (2) Mutator (3) Rule (4) RuleSet (5) CREvent
+'''
+
+# DeltaSet: Use `path` to specify the field that wants to mutate, and `candidates` is a list of candidate
+#           value for the field.
+# Example:  DeltaSet("spec.headGroupSpec.template.spec.containers.0.name", ['ray-head-1', 'ray-head-2', 'ray-head-3'])
+class DeltaSet:
```

Comment:
A DeltaSet is equivalent to a list of jsonpatch replace operations with fixed path parameter and value parameter varying across candidates.

Reply:
TODO: jsonpatch, update DeltaSet definitions

Reply:
Removed the `DeltaSet` abstraction, and used `JsonPatch` instead.